File: entroly/work_graph_repo.py
"""Conservative Git observation adapter for Entroly AI Work Graph.

This module is orchestration-only. It observes durable repository facts and
normalizes them into the schema owned by ``entroly-engine``. It intentionally
never guesses task intent from filenames, commit prose, or branch names.
"""

from __future__ import annotations

import hashlib
import os
import subprocess
import time
from pathlib import Path
from typing import Any
from urllib.parse import urlsplit


class RepositoryDiscoveryError(RuntimeError):
    """Raised when a repository cannot be observed safely."""


def _run_git(
    cwd: Path,
    *args: str,
    timeout: float = 5.0,
    check: bool = True,
) -> str:
    env = os.environ.copy()
    env.update({"GIT_TERMINAL_PROMPT": "0", "LC_ALL": "C", "LANG": "C"})
    try:
        result = subprocess.run(
            ["git", "-C", str(cwd), *args],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=timeout,
            env=env,
            check=False,
        )
    except (OSError, subprocess.TimeoutExpired) as exc:
        raise RepositoryDiscoveryError(f"git {' '.join(args)} failed: {exc}") from exc
    if check and result.returncode != 0:
        detail = result.stderr.strip() or f"exit {result.returncode}"
        raise RepositoryDiscoveryError(f"git {' '.join(args)} failed: {detail}")
    return result.stdout


def _try_git(cwd: Path, *args: str, timeout: float = 5.0) -> str:
    try:
        return _run_git(cwd, *args, timeout=timeout, check=True).strip()
    except RepositoryDiscoveryError:
        return ""


def _normalize_remote(remote: str) -> str:
    value = remote.strip().rstrip("/")
    if not value:
        return ""
    if value.endswith(".git"):
        value = value[:-4]
    if "://" in value:
        parsed = urlsplit(value)
        host = (parsed.hostname or "").lower()
        if parsed.port:
            host = f"{host}:{parsed.port}"
        path = parsed.path.strip("/")
        return f"{host}/{path}" if host and path else ""
    if ":" in value and "@" in value.split(":", 1)[0]:
        lhs, rhs = value.split(":", 1)
        host = lhs.rsplit("@", 1)[-1].lower()
        return f"{host}/{rhs.strip('/')}"
    return ""


def _repository_id(root: Path) -> str:
    remote = _normalize_remote(_try_git(root, "config", "--get", "remote.origin.url"))
    if remote:
        digest = hashlib.sha256(remote.encode("utf-8")).hexdigest()[:32]
        return f"git:{digest}"
    roots = _try_git(root, "rev-list", "--max-parents=0", "HEAD").splitlines()
    if roots:
        material = f"{roots[0]}\0{root.name}".encode("utf-8", "surrogatepass")
        return f"git-root:{hashlib.sha256(material).hexdigest()[:32]}"
    canonical = os.path.normcase(str(root.resolve())).encode("utf-8", "surrogatepass")
    return f"git-local:{hashlib.sha256(canonical).hexdigest()[:32]}"


def _resolve_root(path: str | os.PathLike[str]) -> Path:
    candidate = Path(path).expanduser().resolve()
    if candidate.is_file():
        candidate = candidate.parent
    root = _try_git(candidate, "rev-parse", "--show-toplevel")
    if not root:
        raise RepositoryDiscoveryError(f"not a Git worktree: {candidate}")
    return Path(root).resolve()


def _default_branch(root: Path, override: str | None) -> str:
    if override:
        return override.removeprefix("origin/")
    remote_head = _try_git(root, "symbolic-ref", "--quiet", "--short", "refs/remotes/origin/HEAD")
    if remote_head.startswith("origin/"):
        return remote_head[len("origin/") :]
    for candidate in ("main", "master"):
        if _try_git(root, "rev-parse", "--verify", "--quiet", f"refs/heads/{candidate}"):
            return candidate
    return ""


def _base_ref(root: Path, default_branch: str) -> str:
    if not default_branch:
        return ""
    for candidate in (f"origin/{default_branch}", default_branch):
        if _try_git(root, "rev-parse", "--verify", "--quiet", candidate):
            return candidate
    return ""


def _ahead_behind(root: Path, base_ref: str) -> tuple[int, int]:
    if not base_ref:
        return 0, 0
    output = _try_git(root, "rev-list", "--left-right", "--count", f"{base_ref}...HEAD")
    try:
        behind, ahead = (int(part) for part in output.split())
    except (TypeError, ValueError):
        return 0, 0
    return ahead, behind


def _git_dir(root: Path) -> Path | None:
    value = _try_git(root, "rev-parse", "--absolute-git-dir")
    return Path(value).resolve() if value else None


def _parse_status(root: Path) -> list[dict[str, Any]]:
    raw = _run_git(root, "status", "--porcelain=v1", "-z", "--untracked-files=all")
    tokens = raw.split("\0")
    changes: list[dict[str, Any]] = []
    index = 0
    while index < len(tokens):
        entry = tokens[index]
        index += 1
        if not entry or len(entry) < 3:
            continue
        xy = entry[:2]
        path = entry[3:].replace("\\", "/")
        old_path = ""
        if ("R" in xy or "C" in xy) and index < len(tokens):
            old_path = tokens[index].replace("\\", "/")
            index += 1

        conflicted = "U" in xy or xy in {"DD", "AA"}
        if xy == "??":
            kind = "untracked"
        elif conflicted:
            kind = "unmerged"
        elif "R" in xy:
            kind = "renamed"
        elif "C" in xy:
            kind = "copied"
        elif "D" in xy:
            kind = "deleted"
        elif "A" in xy:
            kind = "added"
        elif "M" in xy or "T" in xy:
            kind = "modified"
        else:
            kind = "unknown"

        changes.append(
            {
                "path": path,
                "kind": kind,
                "staged": xy[0] not in {" ", "?", "!"},
                "conflicted": conflicted,
                "old_path": old_path,
            }
        )
    changes.sort(key=lambda item: (item["path"], item["old_path"], item["kind"]))
    return changes


def _branch_commits(root: Path, base_ref: str, ahead_by: int, max_commits: int) -> list[dict[str, Any]]:
    if ahead_by <= 0 or not base_ref or max_commits <= 0:
        return []
    limit = min(int(max_commits), 100)
    output = _run_git(
        root,
        "log",
        "--no-decorate",
        f"-n{limit}",
        "--format=%H%x00%ct%x00%P%x00%s%x00",
        f"{base_ref}..HEAD",
    )
    fields = output.split("\0")
    commits: list[dict[str, Any]] = []
    for offset in range(0, len(fields) - 3, 4):
        sha, timestamp, parents, subject = fields[offset : offset + 4]
        sha = sha.strip()
        if not sha:
            continue
        try:
            timestamp_ms = int(timestamp.strip()) * 1000
        except ValueError:
            timestamp_ms = 0
        commits.append(
            {
                "sha": sha,
                "subject": subject.strip(),
                "timestamp_ms": timestamp_ms,
                "parent_shas": [item for item in parents.split() if item],
            }
        )
    commits.sort(key=lambda item: (item["timestamp_ms"], item["sha"]))
    return commits


def discover_repository_observation(
    path: str | os.PathLike[str] = ".",
    *,
    agent_id: str = "",
    session_id: str = "",
    task_hint: dict[str, Any] | None = None,
    default_branch: str | None = None,
    max_commits: int = 20,
    observed_at_ms: int | None = None,
) -> dict[str, Any]:
    """Observe durable Git facts without inferring task intent.

    A clean default-branch checkout produces no changes/commits-ahead and therefore
    remains a Work Graph null control: Rust will not invent an unfinished task.
    """

    root = _resolve_root(path)
    branch_name = _try_git(root, "symbolic-ref", "--quiet", "--short", "HEAD")
    head_sha = _try_git(root, "rev-parse", "--verify", "HEAD")
    default = _default_branch(root, default_branch)
    base = _base_ref(root, default)
    ahead, behind = _ahead_behind(root, base)
    git_dir = _git_dir(root)

    merge_in_progress = bool(git_dir and (git_dir / "MERGE_HEAD").exists())
    rebase_in_progress = bool(
        git_dir
        and ((git_dir / "rebase-merge").exists() or (git_dir / "rebase-apply").exists())
    )

    return {
        "repo_id": _repository_id(root),
        "observed_at_ms": int(observed_at_ms if observed_at_ms is not None else time.time() * 1000),
        "repository_label": root.name,
        "agent_id": agent_id,
        "session_id": session_id,
        "task_hint": task_hint,
        "branch": {
            "name": branch_name,
            "head_sha": head_sha,
            "base_ref": base,
            "default_branch": default,
            "ahead_by": ahead,
            "behind_by": behind,
            "merge_in_progress": merge_in_progress,
            "rebase_in_progress": rebase_in_progress,
            "detached": not bool(branch_name),
        },
        "changes": _parse_status(root),
        "commits": _branch_commits(root, base, ahead, max_commits),
        "verifications": [],
        "decisions": [],
        "claims": [],
        "leases": [],
        "model_executions": [],
    }


__all__ = ["RepositoryDiscoveryError", "discover_repository_observation"]

File: entroly-wasm/js/work_graph_repo.js
'use strict';

// Local repository observation only. Work-state inference/trust remains in Rust.
const crypto = require('crypto');
const fs = require('fs');
const os = require('os');
const path = require('path');
const zlib = require('zlib');
const { execFileSync } = require('child_process');

const MAX_GIT_OUTPUT_BYTES = 8 * 1024 * 1024;
const MAX_CHANGES = 16384;
const MAX_COMMITS = 20;
const MAX_CHECKPOINT_BYTES = 32 * 1024 * 1024;

class RepositoryDiscoveryError extends Error {
  constructor(message) {
    super(message);
    this.name = 'RepositoryDiscoveryError';
  }
}

function runGit(cwd, args, { timeout = 5000, check = true } = {}) {
  const commandArgs = [
    '-c', 'core.fsmonitor=false',
    '-c', 'core.untrackedCache=false',
    '-c', 'submodule.recurse=false',
    '-C', cwd,
    ...args,
  ];
  const env = {
    ...process.env,
    GIT_TERMINAL_PROMPT: '0',
    GIT_OPTIONAL_LOCKS: '0',
    GIT_PAGER: 'cat',
    LC_ALL: 'C',
    LANG: 'C',
  };
  try {
    return execFileSync('git', commandArgs, {
      encoding: 'utf8',
      stdio: ['ignore', 'pipe', 'pipe'],
      timeout,
      maxBuffer: MAX_GIT_OUTPUT_BYTES,
      env,
    });
  } catch (error) {
    if (!check) return '';
    const stderr = error && error.stderr ? String(error.stderr).trim() : '';
    throw new RepositoryDiscoveryError(
      `git ${args.join(' ')} failed: ${stderr || (error && error.message) || 'unknown error'}`,
    );
  }
}

function tryGit(cwd, args, options = {}) {
  try { return runGit(cwd, args, options).trim(); }
  catch (_) { return ''; }
}

function normalizeRemote(remote) {
  let value = String(remote || '').trim().replace(/\/$/, '');
  if (!value) return '';
  if (value.endsWith('.git')) value = value.slice(0, -4);
  if (value.includes('://')) {
    try {
      const parsed = new URL(value);
      const host = parsed.port
        ? `${parsed.hostname.toLowerCase()}:${parsed.port}`
        : parsed.hostname.toLowerCase();
      const repoPath = parsed.pathname.replace(/^\/+|\/+$/g, '');
      return host && repoPath ? `${host}/${repoPath}` : '';
    } catch (_) { return ''; }
  }
  const scp = value.match(/^(?:[^@]+@)?([^:]+):(.+)$/);
  if (scp) return `${scp[1].toLowerCase()}/${scp[2].replace(/^\/+|\/+$/g, '')}`;
  return '';
}

function sha32(value) {
  return crypto.createHash('sha256').update(value).digest('hex').slice(0, 32);
}

function resolveRoot(repoPath) {
  let candidate = path.resolve(String(repoPath || '.'));
  try { if (fs.statSync(candidate).isFile()) candidate = path.dirname(candidate); }
  catch (_) {}
  const root = tryGit(candidate, ['rev-parse', '--show-toplevel']);
  if (!root) throw new RepositoryDiscoveryError(`not a Git worktree: ${candidate}`);
  return path.resolve(root);
}

function repositoryId(root) {
  const remote = normalizeRemote(tryGit(root, ['config', '--get', 'remote.origin.url']));
  if (remote) return `git:${sha32(remote)}`;
  const roots = tryGit(root, ['rev-list', '--max-parents=0', 'HEAD'])
    .split(/\r?\n/).filter(Boolean);
  if (roots.length) return `git-root:${sha32(`${roots[0]}\0${path.basename(root)}`)}`;
  return `git-local:${sha32(process.platform === 'win32' ? root.toLowerCase() : root)}`;
}

function discoverRepositoryIdentity(repoPath = '.') {
  const root = resolveRoot(repoPath);
  return { repo_id: repositoryId(root), root };
}

function validateDefaultBranch(root, override) {
  if (!override) return '';
  const name = String(override).replace(/^origin\//, '').trim();
  if (!name || name.startsWith('-') || /[\0\r\n]/.test(name)) {
    throw new RepositoryDiscoveryError(`invalid default branch override: ${JSON.stringify(override)}`);
  }
  try { runGit(root, ['check-ref-format', `refs/heads/${name}`]); }
  catch (_) {
    throw new RepositoryDiscoveryError(`invalid default branch override: ${JSON.stringify(override)}`);
  }
  return name;
}

function defaultBranch(root, override) {
  const explicit = validateDefaultBranch(root, override);
  if (explicit) return explicit;
  const remoteHead = tryGit(root, ['symbolic-ref', '--quiet', '--short', 'refs/remotes/origin/HEAD']);
  if (remoteHead.startsWith('origin/')) return remoteHead.slice('origin/'.length);
  for (const candidate of ['main', 'master']) {
    if (tryGit(root, ['rev-parse', '--verify', '--quiet', `refs/heads/${candidate}`])) return candidate;
  }
  return '';
}

function baseRef(root, defaultName) {
  if (!defaultName) return '';
  for (const candidate of [`refs/remotes/origin/${defaultName}`, `refs/heads/${defaultName}`]) {
    if (tryGit(root, ['rev-parse', '--verify', '--quiet', candidate])) return candidate;
  }
  return '';
}

function aheadBehind(root, base) {
  if (!base) return [0, 0];
  const parts = tryGit(root, ['rev-list', '--left-right', '--count', `${base}...HEAD`]).split(/\s+/);
  if (parts.length < 2) return [0, 0];
  const behind = Number.parseInt(parts[0], 10);
  const ahead = Number.parseInt(parts[1], 10);
  return [Number.isFinite(ahead) ? ahead : 0, Number.isFinite(behind) ? behind : 0];
}

function parseStatus(root) {
  const raw = runGit(root, [
    'status', '--porcelain=v1', '-z', '--untracked-files=all', '--ignore-submodules=all',
  ]);
  const tokens = raw.split('\0');
  const changes = [];
  for (let index = 0; index < tokens.length;) {
    const entry = tokens[index++];
    if (!entry) continue;
    if (entry.length < 3) throw new RepositoryDiscoveryError('malformed porcelain status record');
    const xy = entry.slice(0, 2);
    const repoPath = entry.slice(3).replace(/\\/g, '/');
    let oldPath = '';
    if ((xy.includes('R') || xy.includes('C')) && index < tokens.length) {
      oldPath = tokens[index++].replace(/\\/g, '/');
    }
    const conflicted = xy.includes('U') || xy === 'DD' || xy === 'AA';
    let kind = 'unknown';
    if (xy === '??') kind = 'untracked';
    else if (conflicted) kind = 'unmerged';
    else if (xy.includes('R')) kind = 'renamed';
    else if (xy.includes('C')) kind = 'copied';
    else if (xy.includes('D')) kind = 'deleted';
    else if (xy.includes('A')) kind = 'added';
    else if (xy.includes('M') || xy.includes('T')) kind = 'modified';
    changes.push({
      path: repoPath,
      kind,
      staged: ![' ', '?', '!'].includes(xy[0]),
      conflicted,
      old_path: oldPath,
    });
    if (changes.length > MAX_CHANGES) {
      throw new RepositoryDiscoveryError(
        `repository has more than ${MAX_CHANGES} changed/untracked paths; ` +
        'refusing a partial Work Graph observation',
      );
    }
  }
  changes.sort((a, b) =>
    `${a.path}\0${a.old_path}\0${a.kind}`.localeCompare(`${b.path}\0${b.old_path}\0${b.kind}`));
  return changes;
}

function branchCommits(root, base, ahead, maxCommits) {
  const limit = Number(maxCommits);
  if (!Number.isSafeInteger(limit) || limit < 0 || limit > MAX_COMMITS) {
    throw new RepositoryDiscoveryError(`maxCommits must be a safe integer between 0 and ${MAX_COMMITS}`);
  }
  if (ahead <= 0 || !base || limit === 0) return [];
  const output = runGit(root, [
    'log', '--no-decorate', `-n${limit}`, '--format=%H%x00%ct%x00%P%x00%s%x00', `${base}..HEAD`,
  ]);
  const fields = output.split('\0');
  const commits = [];
  for (let offset = 0; offset + 3 < fields.length; offset += 4) {
    const sha = fields[offset].trim();
    if (!sha) continue;
    const seconds = Number.parseInt(fields[offset + 1].trim(), 10);
    commits.push({
      sha,
      subject: fields[offset + 3].trim(),
      timestamp_ms: Number.isFinite(seconds) ? seconds * 1000 : 0,
      parent_shas: fields[offset + 2].trim().split(/\s+/).filter(Boolean),
      changed_paths: [],
    });
  }
  commits.sort((a, b) => a.timestamp_ms - b.timestamp_ms || a.sha.localeCompare(b.sha));
  return commits;
}

function projectCheckpointCandidates(root, explicitDir = null) {
  if (explicitDir) return [path.resolve(String(explicitDir))];
  if (process.env.ENTROLY_DIR) return [path.resolve(process.env.ENTROLY_DIR)];
  const projectHash = crypto
    .createHash('sha256')
    .update(path.resolve(root), 'utf8')
    .digest('hex')
    .slice(0, 12);
  return [
    path.join(os.homedir(), '.entroly', 'checkpoints', projectHash),
    path.join(os.tmpdir(), 'entroly', 'checkpoints', projectHash),
  ];
}

function existingCheckpointDir(root, explicitDir = null) {
  for (const candidate of projectCheckpointCandidates(root, explicitDir)) {
    try {
      const stat = fs.lstatSync(candidate);
      if (!stat.isSymbolicLink() && stat.isDirectory()) return candidate;
    } catch (_) {}
  }
  return '';
}

function latestCheckpointMetadata(checkpointDir) {
  if (!checkpointDir) return ['', {}];
  const directory = path.resolve(String(checkpointDir));
  let entries;
  try {
    entries = fs.readdirSync(directory)
      .filter(name => /^(?:ckpt_|checkpoint_).+\.json\.gz$/.test(name))
      .map(name => {
        const full = path.join(directory, name);
        return { name, full, stat: fs.statSync(full) };
      })
      .filter(item => item.stat.isFile() && item.stat.size <= MAX_CHECKPOINT_BYTES)
      .sort((a, b) => b.stat.mtimeMs - a.stat.mtimeMs || b.name.localeCompare(a.name));
  } catch (_) { return ['', {}]; }
  for (const item of entries) {
    try {
      const raw = zlib.gunzipSync(fs.readFileSync(item.full), { maxOutputLength: MAX_CHECKPOINT_BYTES });
      const checkpoint = JSON.parse(raw.toString('utf8'));
      const id = String(checkpoint.checkpoint_id || item.name.replace(/\.json\.gz$/, ''));
      const metadata = checkpoint.metadata && typeof checkpoint.metadata === 'object'
        ? checkpoint.metadata
        : checkpoint;
      return [id, metadata];
    } catch (_) {}
  }
  return ['', {}];
}

function discoverRepositoryObservation(repoPath = '.', options = {}) {
  const root = resolveRoot(repoPath);
  const branchName = tryGit(root, ['symbolic-ref', '--quiet', '--short', 'HEAD']);
  const headSha = tryGit(root, ['rev-parse', '--verify', 'HEAD']);
  const defaultName = defaultBranch(root, options.defaultBranch || null);
  const base = baseRef(root, defaultName);
  const [ahead, behind] = aheadBehind(root, base);
  const gitDirText = tryGit(root, ['rev-parse', '--absolute-git-dir']);
  const gitDir = gitDirText ? path.resolve(gitDirText) : null;
  const mergeInProgress = Boolean(gitDir && fs.existsSync(path.join(gitDir, 'MERGE_HEAD')));
  const rebaseInProgress = Boolean(gitDir && (
    fs.existsSync(path.join(gitDir, 'rebase-merge')) ||
    fs.existsSync(path.join(gitDir, 'rebase-apply'))
  ));
  const changes = parseStatus(root);
  const maxCommits = options.maxCommits == null ? MAX_COMMITS : options.maxCommits;
  const commits = branchCommits(root, base, ahead, maxCommits);
  const meaningfulGit = Boolean(changes.length || ahead || mergeInProgress || rebaseInProgress);

  let taskHint = options.taskHint ? { ...options.taskHint } : null;
  const decisions = [];
  const includeCheckpoint = options.includeCheckpoint !== false;
  const checkpointDir = existingCheckpointDir(root, options.checkpointDir || null);
  if (includeCheckpoint && meaningfulGit && checkpointDir) {
    const [checkpointId, metadata] = latestCheckpointMetadata(checkpointDir);
    if (checkpointId) {
      const sourceRef = `checkpoint:${checkpointId}`;
      if (!taskHint) {
        const task = String(metadata.task || '').trim();
        if (task) {
          const step = String(metadata.step || metadata.current_step || '').trim();
          taskHint = {
            task_id: `checkpoint:${checkpointId}`,
            title: task,
            trust: 'observed',
            explicit_status: 'unknown',
            remaining_work: step ? [step] : [],
            source_kind: 'checkpoint',
            source_ref: sourceRef,
          };
        }
      }
      const rawDecisions = Array.isArray(metadata.decisions) ? metadata.decisions.slice(0, 20) : [];
      rawDecisions.forEach((value, index) => {
        const text = String(value || '').trim();
        if (text) decisions.push({
          decision_id: `${checkpointId}:${index}`,
          text,
          source_ref: sourceRef,
          source_kind: 'checkpoint',
          trust: 'observed',
        });
      });
    }
  }

  const observedAt = options.observedAtMs == null ? Date.now() : Number(options.observedAtMs);
  if (!Number.isSafeInteger(observedAt)) {
    throw new RepositoryDiscoveryError('observedAtMs must be a JavaScript-safe integer');
  }

  return {
    repo_id: repositoryId(root),
    observed_at_ms: observedAt,
    repository_label: path.basename(root),
    agent_id: String(options.agentId || ''),
    session_id: String(options.sessionId || ''),
    task_hint: taskHint,
    branch: {
      name: branchName,
      head_sha: headSha,
      base_ref: base,
      default_branch: defaultName,
      ahead_by: ahead,
      behind_by: behind,
      merge_in_progress: mergeInProgress,
      rebase_in_progress: rebaseInProgress,
      detached: !branchName,
    },
    changes,
    commits,
    verifications: [],
    decisions,
    claims: [],
    leases: [],
    model_executions: [],
  };
}

module.exports = {
  RepositoryDiscoveryError,
  discoverRepositoryIdentity,
  discoverRepositoryObservation,
};
